### ISSUE TITLE
Overriding base types?

### ISSUE THREAD

**Issue by ethindp** [ThePhD/sol2#1490]
I'd like to override the base types of Lua (e.g. maybe replacing the integer type with an arbitrary-precision integer, or a string with a custom string class). Is this even possible? I mean, I know I could register a usertype called String or something, but then if I used functions in my Lua code that used that custom type I'd have to do:

```lua
local var = func(custom_string("test"))
```

Or similar. That isn't ideal. I could make conversion transparent, but then if I (say) wanted to expose methods in QString I'd need to write hundreds of lambdas to convert back and forth. Is this just a pipe dream, or is it doable?

**Comment by ethindp** [comment]
@Rochet2 Yep, exactly what I mean. (As an aside, it doesn't appear that Sol plays well with nested STL types like `std::set<std::variant<...>>`... any advice for handling that? I'm asking because I want to expose a set that only contains two possible types, but I'm honestly unsure how to register variants in Lua. It would seem kinda awkward having to call `variant:new(...)` or something. I am using string_views, so I don't know if Sol knows how to work with those.)

**Comment by ethindp** [comment]
@Rochet2 As an alternative to customizing via custom converters, could I just "extend" the existing string type? It would be really cool to be able to do `"Hi":index_of(...)` or something. If not, I can use the resources you listed, but thought I'd ask anyway.

**Comment by ethindp** [comment]
But for "replacing" existing types I'll definitely look into what you said -- it would be nice having an arbitrary-precision integer/real type.

**Comment by Rochet2** [comment]
> It would be really cool to be able to do `"Hi":index_of(...)` or something

You can do it like this

```lua
local foo = "A"
function string.foo(s) return s..s end -- dummy example
print(foo:foo())
```

> As an aside, it doesn't appear that Sol plays well with nested STL types like `std::set<std::variant<...>>`... any advice for handling that?

I dont really have anything for that. You can usually wrap things into structs for example to avoid issues.
